src/WorkflowCore/Services/DefinitionStorage/DefinitionLoader.cs

```diff
@@ -173,12 +173,15 @@ private void AttachInputs(StepSourceV1 source, Type dataType, Type stepType, Wor
                 var dataParameter = Expression.Parameter(dataType, "data");
                 var contextParameter = Expression.Parameter(typeof(IStepExecutionContext), "context");
                 var sourceExpr = DynamicExpressionParser.ParseLambda(new [] { dataParameter, contextParameter }, typeof(object), input.Value);
-                var targetExpr = Expression.Property(Expression.Parameter(stepType), input.Key);
 
-                step.Inputs.Add(new DataMapping()
+                var stepParameter = Expression.Parameter(stepType, "step");
+                var targetProperty = Expression.Property(stepParameter, input.Key);
+                var targetExpr = Expression.Lambda(targetProperty, stepParameter);
+
+                step.Inputs.Add(new DataMapping
                 {
                     Source = sourceExpr,
-                    Target = Expression.Lambda(targetExpr)
+                    Target = targetExpr
```

Comment:
Could we add a scenario test the exercises this code with an actual JSON defined workflow that does this?
Would you also be willing to write some unit tests around the `DefinitionLoader` class? (I know we don't have any yet, but that would be awesome)

Reply:
Sure! There is already `WorkflowCore.UnitTests.Services.DefinitionStorage.DefinitionLoaderTests` Should I add them there?

Reply:
I added a new unittest that loads a workflow that uses a dynamic datatype, I had to change the conditions a little bit to make it works with the indexer returning object instead of just int or just bool.